File: app/src/main/java/com/udangtangtang/haveibeen/util/GeocodingHelper.java
package com.udangtangtang.haveibeen.util;

import android.content.Context;
import android.location.Address;
import android.location.Geocoder;
import android.widget.Toast;

import java.io.IOException;
import java.util.List;

public class GeocodingHelper {
    private Context context;
    private Geocoder geocoder;
    private List<Address> addressList;

    // AdminArea -> 특별, 광역시/도
    // Locality -> 시
    // SubLocality -> 구 (특별 / 광역시의 구 포함)
    // Thoroughfare -> 읍/면/동/로
    private String[] address;

    public GeocodingHelper(Context mContext, double latitude, double longtitude) {
        context = mContext;

        // 역 지오코딩
        geocoder = new Geocoder(context);
        address=new String[4];
        addressList = null;
        try {
            // Geocoder를 통해 주소 획득
            addressList = geocoder.getFromLocation(latitude, longtitude, 10);
        } catch (
                IOException e) {
        }
        if (addressList != null) {
            if (addressList.size() == 0) {
                Toast.makeText(context, "해당 지역의 주소를 제공할 수 없습니다.", Toast.LENGTH_LONG).show();
            } else {
                address[0]=addressList.get(0).getAdminArea();
                address[1]=addressList.get(0).getLocality();
                address[2]=addressList.get(0).getSubLocality();
                address[3]=addressList.get(0).getThoroughfare();
            }
        }
    }



    public String getAddress() {
        String result="";
        for(int i=0; i<4; i++){
            // null을 건너 뛰고 각 단위를 이어 붙임
            // ex) 경기도 광명시 null 일직동 => 경기도 광명시 일직동
            //     경기도 안양시 만안구 석수2동 과 달리 구 단위가 없기 때문
            if(address[i]!=null){
                result+=address[i];
                if(i!=3){ result+=" "; }
            }
        }
        return result;
    }
}

File: app/src/main/java/com/udangtangtang/haveibeen/util/GeocodingHelper.kt
package com.udangtangtang.haveibeen.util

import android.app.Application
import android.content.*
import android.icu.text.AlphabeticIndex.Record
import android.location.Geocoder
import android.widget.Toast
import android.location.Address
import android.location.Geocoder.GeocodeListener
import android.os.Build
import android.util.Log
import androidx.annotation.RequiresApi
import androidx.viewpager2.widget.ViewPager2
import com.bumptech.glide.Glide.init
import com.udangtangtang.haveibeen.repository.RecordRepository
import com.udangtangtang.haveibeen.util.PictureScanHelper
import kotlinx.coroutines.*
import kotlinx.coroutines.Dispatchers.IO
import java.io.IOException

@RequiresApi(Build.VERSION_CODES.TIRAMISU)
class GeocodingHelper(private val context: Context, private val db : RecordRepository) {
    // TODO : 주소 수정 기능
    private var geocoder: Geocoder
    var result=""

    // AdminArea -> 특별, 광역시/도
    // Locality -> 시
    // SubLocality -> 구 (특별 / 광역시의 구 포함)
    // Thoroughfare -> 읍/면/동/로
    // SubThoroughfare -> 번지

    init {
        // 역 지오코딩
        geocoder = Geocoder(context)
    }

    companion object{
        private val TAG="GeocodingHelper"

        fun getAddressToString(addr:Address):String{
            var address=""
            with(addr){
                Log.d(TAG,">>>>>>>>>>>> :"+"admin : "+adminArea+" subadmin: "+subAdminArea+" locality: "+locality+" sublocality : "+subLocality+" thoroughfare : "+thoroughfare+" subthoroughfare : "+subThoroughfare+" premises : "+premises+" feature : "+featureName)
                if(!adminArea.isNullOrEmpty()){
                    address+=adminArea
                    address+=" "
                }
                if(!locality.isNullOrEmpty()){
                    address+=locality
                    address+=" "
                }
                if(!subLocality.isNullOrEmpty()){
                    address+=subLocality
                    address+=" "
                }
                if(!thoroughfare.isNullOrEmpty()){
                    address+=thoroughfare
                    address+=" "
                }
                if(!subThoroughfare.isNullOrEmpty()) {
                    address += subThoroughfare
                }
            }
            return address
        }
    }


    @RequiresApi(Build.VERSION_CODES.TIRAMISU)
    fun getAddress(latitude: Double, longtitude: Double) {
        // Geocoder를 통해 주소 획득
        geocoder.getFromLocation(
            latitude,
            longtitude,
            3,
            object : Geocoder.GeocodeListener {
                override fun onGeocode(addressList: MutableList<Address>) {
                        Log.d(TAG, addressList[0].toString())
                    CoroutineScope(Dispatchers.IO).launch {
                        db.updatePictureAddress(latitude, longtitude, getAddressToString(addressList[0]))
                    }
                }
                override fun onError(errorMessage: String?) {
                    super.onError(errorMessage)
                }
            })
        }


    }